### ISSUE TITLE
GAP crashes when returning from 'method no found' errors.

### ISSUE THREAD

**Issue by ChrisJefferson** [gap-system/gap#2447]
Example:

```
gap> f := function( a ) return a + a; end;

function( a ) ... end
gap> f( () );
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `+' on 2 arguments at /mnt/c/Users/caj/files/reps/gap/gap/lib/methsel2.g:250 called from

a + a at *stdin*:2 called from
<function "f">( <arguments> )
 called from read-eval loop at *stdin*:3
you can 'quit;' to quit to outer loop, or you can 'return;' to continue
brk> return;
<segfault>
```

The problem seems to be around opers.c:2178, where we call HandleMethodNotFound, and don't check the return value isn't 0. In fact, I think (from my brief usage), it's impossible to actually return a function from CallHandleMethodNotFound, so this should instead just error out, rather than try to loop (but that should be checked).

Git blame suggests @stevelinton was last around this code, but it could be a combination of various function related changes. It did work in 4.8.

**Comment by fingolfin** [comment]
This regression affects the stable-4.9 branch. We may wish to put a warning into the GAP 4.9.1 release notes (@alex-konovalov can we still do that, or is it too late?), and work on a fix for GAP 4.9.2, and release that ASAP after GAP 4.9.1

**Comment by fingolfin** [comment]
While @stevelinton indeed performed a major rewrite of the associated code, I also mucked around with it recently. I think it would be best to perform a `git bisect` to see which commit exactly introduced it, then think about how we can resolve it.

But it is indeed odd that we are checking for a return value from `HANDLE_METHOD_NOT_FOUND`, when that function doesn't return anything (so the return value is always a zero pointer on the GAP level). Huh. Perhaps this used to be different in the distant past? E.g. one might conceive of somebody doing an `InstallMethod` on that break prompt, then resume... But that's not going to work in the current setup either. Odd.

**Comment by fingolfin** [comment]
`git bisect` reveals that this regression was introduced by 9391a8ab1713d70964496c9b0a737ae32fa16829 i.e. the refactoring @stevelinton did. Ah well, that's always the risk with a good refactoring, but I still think it was 100% worth it; we'll just have to iron out this kink.

**Comment by fingolfin** [comment]
Fixed in PR #2449 with a minimal change, so that we can easily backport to stable-4.9

However, as @ChrisJefferson points out, returning from that "method not found" error / break loop actually is not helpful, as you are not allowed to return an actual function; and it has been like that even in GAP 4.4, I just tested it there. Several ways to deal with this come to mind:

1. we could add an `ErrorReturnObj` on the library level, and use that in `HANDLE_METHOD_NOT_FOUND`, instead of `Error`; with a prompt telling the user they can return an alternate function.
2. We could drop that return value, i.e. change `CallHandleMethodNotFound` to not return, and change  `HANDLE_METHOD_NOT_FOUND` to end with `ErrorNoReturn` instead of `Error`. (As a safeguard, we might also add a `ErrorQuit` or a `GAP_ASSERT` at the end of `CallHandleMethodNotFound`).

I am in favor of the second option, because (a) returning a method wasn't actually supported before; (b) it is easier to implement and test, (c) I actually think allowing the user to return a custom function there is a bad idea. The only argument in favor I can think of is: "but what if I had this computation running for hours/days/weeks, and only at the end it got stuck due to this tiny typo, which I could work around if you let me return a custom function". But this goes completely against reproducibility of scientific results! You should re-run your computation after the fix anyway (and think about ways to store partial results in a way that makes stopping and resuming it feasible), not use a non-reproducible, error prone hotfix like that.
